### ISSUE TITLE
The computer vision v1 spec lists regions hosts, but v2 does not

### ISSUE THREAD

**Issue by pimterry** [Azure/azure-rest-api-specs#5534]
The [v1 spec for computer vision](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/ComputerVision/stable/v1.0/ComputerVision.json#L20-L27) includes an `x-ms-parameterized-host` section, which defines the host as `{AzureRegion}.api.cognitive.microsoft.com`, referencing an AzureRegion param that lists all valid regions, and thereby clearly describing all available hosts for this API.

This is great, and very helpful.

The [v2 spec](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/ComputerVision/stable/v2.0/ComputerVision.json#L20-L28) doesn't do this.

Instead it defines a `{Endpoint}` parameterized host, and then [defines the Endpoint param](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/ComputerVision/stable/v2.0/ComputerVision.json#L1277-L1285) as any valid string, which isn't useful at all (and isn't really correct either: there any many strings that aren't valid hosts for this API).

Is there a specific reason behind this? If not, it would be very helpful to have the full list of valid hosts available on this spec, as it is for v1. This probably applies to other specs too, this is just one case I spotted.

**Comment by cthrash** [comment]
There were two reasons for which this was changed, both having to do with that the simple AzureRegion model no longer worked:
1) Some Azure regions having a naming scheme that does not use the same pattern, and
1) We want to make the client SDKs usable with [containers](https://docs.microsoft.com/en-us/azure/cognitive-services/cognitive-services-container-support), where the endpoint name will *definitely* not follow the pattern.

To make this slightly less confusing, the `Overview` blade of the Azure Portal contains the string you should use during client instantiation (see below).

![image](https://user-images.githubusercontent.com/14881441/60295043-a0856c80-98d7-11e9-983e-064ead18d839.png)

**Comment by pimterry** [comment]
@cthrash It's useful that that info is available in the portal, but that doesn't help me I'm afraid. I'm using the spec as an input to other OpenAPI tools, and as part of that I need a machine-readable definition of the common endpoints. It doesn't need to be 100% complete, but knowing the typical format would be very useful, and the previous spec did that quite well.

Is there no way to include this info in the spec? It sounds like the region list is no longer complete, but is still _often_ correct. Would it be possible to include the region URL format there, and _also_ indicate the other potential formats in addition?

**Comment by cthrash** [comment]
I think relying on a static list is problematic, since not all services are deployed in all the [regions](https://azure.microsoft.com/en-us/global-infrastructure/services/?products=cognitive-services), and the [list](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/Common/ExtendedRegions.json) is not even comprehensive.

I don't know precisely how it's done, but when you provision any give Cognitive Services service through the Azure Portal, you will see the current list of supported regions, so can your tool plug in to that?

**Comment by pimterry** [comment]
> I don't know precisely how it's done, but when you provision any give Cognitive Services service through the Azure Portal, you will see the current list of supported regions, so can your tool plug in to that?

@cthrash Sadly I can't. I'm inspecting live HTTP traffic, and matching that against OpenAPI specs to work out which services it's talking to, so I can overlay metadata, docs & request/response validation to help development (see [HttpToolkit.tech](https://httptoolkit.tech)). Many of my users are Azure customers, but I'm not directly involved in my users' Azure configuration at all (and I don't want to be).

For most other services in [the OpenAPI directory](https://github.com/APIs-guru/openapi-directory) the specs alone are easily sufficient for that, as they have one or two specific endpoints that are simple to recognize. For similar services like AWS too it's not too bad, as the list of regions is well publicized, with only a few exceptions to manually handle. E.g. check out the [S3 spec](https://github.com/APIs-guru/openapi-directory/blob/master/APIs/amazonaws.com/s3/2006-03-01/swagger.yaml#L41-L143) for a complicated example ([generated from the AWS SDK + some basic rules](https://github.com/APIs-guru/aws2openapi/blob/master/index.js#L76-L131)).

For Azure though it seems to be much more challenging, since the set of endpoints available for each service doesn't seem to be formally documented anywhere at all...

Is there a machine-readable format of that region/product matrix you listed there available? What's the reason why the JSON extended regions list you linked isn't comprehensive?

In my dream world, the information in that matrix would be encoded in the specs for each of the services, including the service/region URL format, and potentially a field signifying that _any_ URL might be used (for the container scenario). That could then easily be used to generate that webpage list. I know that's a big ask of course, just outlining the kind of data I'm searching for. Data like this would make development like mine that integrates with the specifications directly much easier & more effective.

**Comment by pimterry** [comment]
@yangyuan that's true, but the data of some endpoints which are _definitely_ cognitive services is also useful information!

It's helpful to be able to find out which endpoints I might be able to send requests to, and it's invaluable for recognizing what service a given URL is for (my case). I agree you need a wildcard too, but having some structured list of known endpoints somewhere would be much better than the current solution.

This data definitely exists, it's even listed in your docs: https://westus.dev.cognitive.microsoft.com/docs/services/5adf991815e1060e6355ad44/operations/56f91f2e778daf14a499e1fa (which is great, that's very useful). It's just not included in the structured data here for some reason. Right now, for the OpenAPI directory I'm going to have to scrape those docs and add the URLs to this spec. That doesn't seem like a good result.
